base/serialize.jl

```diff
@@ -362,13 +328,50 @@ function serialize(s::SerializationState, t::Task)
     end
 end
 
-function serialize_type_data(s, t)
-    tname = t.name.name
-    serialize(s, tname)
-    mod = t.name.module
-    serialize(s, mod)
+function serialize(s::SerializationState, t::TypeName)
+    serialize_cycle(s, t) && return
+    writetag(s.io, TYPENAME_TAG)
+    serialize(s, t.name)
+    serialize(s, t.module)
+    serialize(s, t.names)
+    serialize(s, t.primary.super)
+    serialize(s, t.primary.parameters)
+    serialize(s, t.primary.types)
+    serialize(s, t.primary.size)
+    serialize(s, t.primary.abstract)
+    serialize(s, t.primary.mutable)
+    serialize(s, t.primary.ninitialized)
+    if isdefined(t, :mt)
+        serialize(s, t.mt)
+    else
+        writetag(s.io, UNDEFREF_TAG)
+    end
+end
+
+# decide whether to send all data for a type (instead of just its name)
+function should_send_whole_type(s, t::ANY)
+    # TODO improve somehow
+    return t.name.module === Main
+end
+
+function serialize_type_data(s, t::ANY, type_itself::Bool)
+    whole = should_send_whole_type(s, t)
+    form = type_itself ? UInt8(0) : UInt8(1)
+    if whole
+        form |= UInt8(2)
+    end
+    writetag(s.io, DATATYPE_TAG)
+    write(s.io, form)
```

Comment:
Since there's only three of these does it make sense to have different tags? Are you anticipating the form to expand in the future?